### PR TITLE
RequestMiddleware

### PR DESCRIPTION
**Pull request by aniketfuryrocks**
Implementation for #1095.

### CODE REVIEW

File: server/src/middleware/request_middleware.rs [hunk 1]
```diff
@@ -0,0 +1,92 @@
...
+	error::Error,
+	future::Future,
+	pin::Pin,
+	sync::Arc,
+	task::{Context, Poll},
+};
+use tower::{Layer, Service};
+
+/// Action undertaken by a middleware.
+pub enum RequestMiddlewareAction {
```
Comment (aniketfuryrocks) [inline server/src/middleware/request_middleware.rs:11]: The clippy policy requires this to be debug. But it's not possible due to pin.

File: server/src/middleware/request_middleware.rs [hunk 2]
```diff
@@ -0,0 +1,101 @@
...
+impl<F> RequestMiddleware for F
+where
+	F: Fn(hyper::Request<hyper::Body>) -> RequestMiddlewareAction + Sync + Send + 'static,
+{
+	fn on_request(&self, request: hyper::Request<hyper::Body>) -> RequestMiddlewareAction {
+		(*self)(request)
+	}
+}
+
+/// Layer that applies [`RequestMiddlewareService`] which proxies the `GET /path` requests to
```
[the remark this answers is not in the document]
Comment (niklasad1) [reply server/src/middleware/request_middleware.rs:45]: not just `GET /path` if I understand this correctly

File: server/src/middleware/mod.rs [hunk 3]
```diff
@@ -2,3 +2,5 @@
 
 /// Proxy `GET /path` to internal RPC methods.
 pub mod proxy_get_request;
+/// Proxy switch requset
```
Comment (niklasad1) [inline server/src/middleware/mod.rs:5]: I think `request_middleware` is too vague and might confuse users.

Perhaps something like a `proxy_http_request` and write some better documentation that it 
exposes the HTTP request and it's possible to override the default jsonrpsee handler response.

File: examples/examples/request_middleware.rs [hunk 4]
```diff
@@ -0,0 +1,87 @@
...
+		ServerBuilder,
+	},
+	types::ErrorObject,
+};
+use std::net::SocketAddr;
+
+#[rpc(server)]
+pub trait MyRpc {
+	#[method(name = "sayHi")]
+	fn say_hi(&self) -> Result<String, ErrorObject<'static>>;
```
Comment (niklasad1) [inline examples/examples/request_middleware.rs:15]:
```suggestion
	fn say_hi(&self) -> String;
```

File: examples/examples/request_middleware.rs [hunk 5]
```diff
@@ -0,0 +1,87 @@
...
+}
+
+pub struct MyRpc;
+
+impl MyRpcServer for MyRpc {
+	fn say_hi(&self) -> Result<String, ErrorObject<'static>> {
```
Comment (niklasad1) [inline examples/examples/request_middleware.rs:21]:
```suggestion
	fn say_hi(&self) -> String {
```

**Review by niklasad1 — COMMENTED** [review]
Thanks for the PR.

The idea is that users can just write their own middleware and but this more generic
than our existing `proxy HTTP get request` so I'm good with it general.

We could even base the `proxy HTTP get request` on top in the middleware you introduced in this PR.

But it's quite hard to understand the difference between `proxy_get_request` and `request_middleware`, so to have this in better documentation is required.

//cc @lexnv @jsdw thoughts?

**Comment by niklasad1** [comment]
We have JSON-RPC specific call middleware which provide something similar and works for both WebSocket and HTTP.

**Pull request closed without merging**